src/main/java/org/alfresco/repo/dictionary/DictionaryModelType.java

```diff
@@ -434,7 +475,60 @@ public void onCreateNode(ChildAssociationRef childAssocRef)
                 queueModel(nodeRef);
             }
         }
-    }
+    }
+    
+    /**
+     * Validate that the definition node is a child of the correct model
+     * location node, e.g. "/Company Home/Data Dictionary/Models"
+     */
+    private boolean isValidLocation(NodeRef definitionNode)
+    {
+        StoreRef storeRef = repositoryModelsLocation.getStoreRef();
+        NodeRef rootNode = nodeService.getRootNode(storeRef);
+        List<NodeRef> nodeRefs = searchService.selectNodes(rootNode, repositoryModelsLocation.getPath(), null, namespaceService, false);
+
+        if (nodeRefs.isEmpty() || nodeRefs.size() > 1)
+        {
+            throw new IllegalStateException(
+                    "Incorrect number of nodes (" + nodeRefs.size() + ")" + " found for workflow location: " + repositoryModelsLocation.getPath());
+        }
+
+        NodeRef modelParent = nodeRefs.get(0);
+
+        for (ChildAssociationRef assoc : nodeService.getParentAssocs(definitionNode))
+        {
+            if (assoc.getParentRef().equals(modelParent))
+            {
+                // The model definition is contained in the correct location
+                return true;
+            }
+        }
+
+        // Invalid location
+        return false;
+    }
+    
+    private boolean isValidLocationOfParentNode(NodeRef parentRef){
+        StoreRef storeRef = repositoryModelsLocation.getStoreRef();
+        NodeRef rootNode = nodeService.getRootNode(storeRef);
+        List<NodeRef> nodeRefs = searchService.selectNodes(rootNode, repositoryModelsLocation.getPath(), null, namespaceService, false);
+
+        if (nodeRefs.isEmpty() || nodeRefs.size() > 1)
+        {
+            throw new IllegalStateException(
+                    "Incorrect number of nodes (" + nodeRefs.size() + ")" + " found for workflow location: " + repositoryModelsLocation.getPath());
+        }
+
+        NodeRef modelParent = nodeRefs.get(0);
```

Comment:
We have the same code here as in "isValidLocation()". We could extract this code into a private method "private NodeRef getModelsLocation()" or so.